### PR TITLE
Part of #11159: Purge usage of malloc() and free() in FileScanner.cpp

### PR DESCRIPTION
**Pull request by frutiemax**

### CODE REVIEW

File: src/openrct2/core/FileScanner.cpp [hunk 1]
```diff
@@ -73,8 +73,8 @@ class FileScannerBase : public IFileScanner
     std::stack<DirectoryState> _directoryStack;
 
     // Current
-    FileInfo* _currentFileInfo;
-    utf8* _currentPath;
+    std::unique_ptr<FileInfo> _currentFileInfo;
```
Comment (ZehMatt) [inline src/openrct2/core/FileScanner.cpp:76]: This allocation can be avoided by having it a basic member of FileScanner.

File: src/openrct2/core/FileScanner.cpp [hunk 2]
```diff
@@ -83,32 +83,29 @@ class FileScannerBase : public IFileScanner
         _recurse = recurse;
         _patterns = GetPatterns(Path::GetFileName(pattern));
 
-        _currentPath = Memory::Allocate<utf8>(MAX_PATH);
-        _currentFileInfo = Memory::Allocate<FileInfo>();
+        _currentPath.resize(MAX_PATH);
+        _currentFileInfo = std::make_unique<FileInfo>();
```
Comment (ZehMatt) [inline src/openrct2/core/FileScanner.cpp:87]:
```suggestion
        _currentFileInfo = FileInfo{};
```
No need to have unique_ptr

File: src/openrct2/core/FileScanner.cpp [hunk 3]
```diff
@@ -83,32 +83,29 @@ class FileScannerBase : public IFileScanner
...
     ~FileScannerBase() override
     {
-        Memory::Free(_currentPath);
-        Memory::Free(_currentFileInfo);
     }
 
     const FileInfo* GetFileInfo() const override
     {
-        return _currentFileInfo;
+        return _currentFileInfo.get();
```
Comment (ZehMatt) [inline src/openrct2/core/FileScanner.cpp:98]: Should return const FIleInfo&

File: src/openrct2/core/FileScanner.cpp [hunk 4]
```diff
@@ -83,32 +83,28 @@ class FileScannerBase : public IFileScanner
         _recurse = recurse;
         _patterns = GetPatterns(Path::GetFileName(pattern));
 
-        _currentPath = Memory::Allocate<utf8>(MAX_PATH);
-        _currentFileInfo = Memory::Allocate<FileInfo>();
+        _currentPath.resize(MAX_PATH);
```
Comment (ZehMatt) [inline src/openrct2/core/FileScanner.cpp:86]: This line can be dropped.

File: src/openrct2/core/FileScanner.cpp [hunk 5]
```diff
@@ -83,32 +83,28 @@ class FileScannerBase : public IFileScanner
...
 
         Reset();
     }
 
     ~FileScannerBase() override
     {
-        Memory::Free(_currentPath);
-        Memory::Free(_currentFileInfo);
     }
```
Comment (tupaschoal) [inline src/openrct2/core/FileScanner.cpp:93]: If the `dtor` will become empty, make it: 
```c++
~FileScannerBase() override = default;
```

File: src/openrct2/core/FileScanner.cpp [hunk 6]
```diff
@@ -83,32 +83,28 @@ class FileScannerBase : public IFileScanner
...
 
     const FileInfo* GetFileInfo() const override
```
Comment (tupaschoal) [inline src/openrct2/core/FileScanner.cpp:95]: Didn't @ZehMatt  mean:
```suggestion
    const FileInfo& GetFileInfo() const override
```

File: src/openrct2/core/FileScanner.cpp [hunk 7]
```diff
@@ -83,32 +83,28 @@ class FileScannerBase : public IFileScanner
...
     {
-        return _currentPath;
+        return _currentPath.c_str();
     }
 
     const utf8* GetPathRelative() const override
     {
-        // +1 to remove the path separator
-        return _currentPath + _rootPath.size() + 1;
+        return _currentPath.substr(_rootPath.size()).c_str();
```
Comment (tupaschoal) [inline src/openrct2/core/FileScanner.cpp:107]: This is problematic, because it is returning a pointer to a temporary

**Comment by IntelOrca** [comment]
@frutiemax why did you close this?

**Pull request closed without merging**